### PR TITLE
feat: Immediate file visibility for new integrated objects  (#781)

### PR DESCRIPTION
**Pull request by NoopDog**
The goal of this PR  was to get integrated objects and soon source datasets to display immediately in the tracker UI after a file record is received from the S3 notification service and the page is refreshed. We are there for component atlases, but source datasets are created ok, but not read properly (no join yet to the files table).

The file IDs are still the IDs for the source datasets and component atlases. However, we create the associated metadata objects (source dataset and component atlas) immediately when we receive a file, and also update the sd_info or component_info to {} when we receive an update for these objects. These fields will be reset once the validation happens.

The aggregation of source dataset info on the linked component atlas is halfway removed. We will update this to collect the aggregates for the source datasets, as well as the total that made it into the atlas.

There are some changes to tests:

1. A lot of new tests as I moved to unit testing
2. I had trouble with tests releasing resources (db connections), so you see some work on test cleanup.
3. I had to mock KY. For some reason, the HTTP mocks were not being applied to specific tests, and the tests were throwing a "not ready" exception, indicating that the cache had not yet been loaded from CELLxGENE.
4. I made the mock that sets the test database in the test setup (and left it in all the other testss) I had a case where I forgot to setup the mock and my dev database got updated to the test database. 

# Database changes

From migration: [migrations/1756223678377_add-component-atlas-id-to-files.ts](cci:7://file:///Users/dave/projects/hca-atlas-tracker/migrations/1756223678377_add-component-atlas-id-to-files.ts:0:0-0:0)

- **Dropped**
  - Constraint: `fk_files_atlas_id` on `hat.files`
  - Column: `atlas_id` from `hat.files`

- **Added columns** (table `hat.files`)
  - `component_atlas_id uuid NULL` — FK target to `hat.component_atlases(id)`
  - `source_dataset_id uuid NULL` — FK target to `hat.source_datasets(id)`
  - `sns_message_id varchar(255) NOT NULL` — for SNS idempotency

- **Added constraints**
  - `fk_files_component_atlas_id` → `hat.files(component_atlas_id)` refs `hat.component_atlases(id)` ON UPDATE CASCADE ON DELETE SET NULL
  - `fk_files_source_dataset_id` → `hat.files(source_dataset_id)` refs `hat.source_datasets(id)` ON UPDATE CASCADE ON DELETE SET NULL
  - `uq_files_sns_message_id` → UNIQUE on `hat.files(sns_message_id)`
  - Replaced check constraint `ck_files_exclusive_parent_relationship` on `hat.files` to enforce:
    - If `file_type='source_dataset'`: `source_dataset_id` NOT NULL and `component_atlas_id` NULL
    - If `file_type='integrated_object'`: `component_atlas_id` NOT NULL and `source_dataset_id` NULL
    - If `file_type='ingest_manifest'`: both `source_dataset_id` and `component_atlas_id` NULL

- **Added indexes** (table `hat.files`)
  - Index on `component_atlas_id`
  - Index on `source_dataset_id`

- **Altered columns** (table `hat.source_datasets`)
  - `source_study_id` made nullable (removed NOT NULL)

# From CoPilot

This pull request primarily adds new and improved tests for atlas lookup and version normalization, and updates several test suites for better reliability and consistency. It also introduces a new mock for the `ky` HTTP client and improves database connection handling in tests. These changes enhance the robustness of the test suite and ensure better coverage of edge cases.

### New and Improved Tests

* Added comprehensive tests for `getAtlasByNetworkVersionAndShortName`, covering successful lookups, error cases, and version-matching edge cases in `__tests__/atlas-lookup.test.ts`.
* Added parameterized tests for `normalizeAtlasVersion` in `__tests__/version-normalization.test.ts`, including valid mappings and expected failures for malformed versions.

### Test Reliability and Consistency

* Updated refresh-related tests (e.g., `__tests__/api-refresh.test.ts`, `__tests__/revalidate-on-refresh.test.ts`) to account for auto-start being disabled, resulting in changes to expected call counts and explicit triggering of refresh logic. [[1]](diffhunk://#diff-ee608352b1b4186b02c4245510d34825c8f827948da38904bc05c0d0e5e79876L138-R144) [[2]](diffhunk://#diff-ee608352b1b4186b02c4245510d34825c8f827948da38904bc05c0d0e5e79876L188-R213) [[3]](diffhunk://#diff-f4900ea657d5c934bdd904a5b3261a33463066f267ea98b5c7fcc8e730ce6bf2L87-R93) [[4]](diffhunk://#diff-f4900ea657d5c934bdd904a5b3261a33463066f267ea98b5c7fcc8e730ce6bf2L105-R116)
* Improved database test teardown by ensuring `endPgPool` is awaited in several test files, reducing the risk of unclosed connections. [[1]](diffhunk://#diff-3a0d0dbe901f8a786b5c3dbc8a53820e692e401070ff19890669f3fb710d0941L30-R30) [[2]](diffhunk://#diff-5a5b61aebebcf36f8af9bf3f42a9ecbbfa7dead30e794e2f3ee11c7b7aff84b9L170-R171) [[3]](diffhunk://#diff-aefb5251b0d3786d8a88534da749c805528d7d0ee72a75a4ee7fd750475a8001L342-L354)

### Mocking and Utilities

* Introduced a new Jest mock for the `ky` HTTP client in `__mocks__/ky.ts`, with helper functions for controlling mock responses in tests.

### Test Coverage for Service Methods

* Added tests for `createComponentAtlas` and `updateComponentAtlas` service functions, including edge cases for non-existent resources in `__tests__/api-atlases-id-component-atlases.test.ts`.

### Minor Test Refactors

* Refactored calls to `getSourceDataset` in relevant test files to use an object parameter for improved clarity and flexibility. [[1]](diffhunk://#diff-033501d03dd2199d66eaf6396a9a4a7ce88db90271fe16a1316d97b231f9f417L225-L233) [[2]](diffhunk://#diff-033501d03dd2199d66eaf6396a9a4a7ce88db90271fe16a1316d97b231f9f417L267-R268) [[3]](diffhunk://#diff-60dfc94659247711f29d35ed6d106a412cd0e400b9281f8af5b8949f2650e580L385-R389) [[4]](diffhunk://#diff-60dfc94659247711f29d35ed6d106a412cd0e400b9281f8af5b8949f2650e580L415-R419)
* Updated import statements and variable declarations for clarity and type safety in test files. [[1]](diffhunk://#diff-aefb5251b0d3786d8a88534da749c805528d7d0ee72a75a4ee7fd750475a8001L10-R10) [[2]](diffhunk://#diff-aefb5251b0d3786d8a88534da749c805528d7d0ee72a75a4ee7fd750475a8001R29-R30)

These updates collectively improve test reliability, coverage of edge cases, and maintainability.

### CODE REVIEW

**Review by hunterckx — CHANGES_REQUESTED** [review]
Misc:
- I think the tests for the API routes should be limited to testing the API handlers and should note include more unit-level tests

File: app/services/s3-notification.ts [hunk 1]
```diff
@@ -139,74 +155,215 @@ function convertS3VersionToDbVersion(s3Version: string): string {
...
+async function createIntegratedObject(
+  atlasId: string | null,
+  s3Object: { eTag: string; key: string },
+  _: string | null,
+  transaction: PoolClient
+): Promise<string> {
+  const title = getTitleFromS3Key(s3Object.key);
+
+  const componentAtlas = await createComponentAtlas(
+    atlasId!,
```
[the remark this answers is not in the document]
Comment (hunterckx) [reply app/services/s3-notification.ts:215]: Poked around a bit and realized that `atlasId` does not need to be nullable in the first place

**Comment by hunterckx** [comment]
Is it okay for the migration to require manually deleting existing files? (Currently it fails if there are any files because the new SNS message ID column has no values but is required to be non-null)

**Comment by NoopDog** [comment]
Yes we will delete files before migration.
Thx!

Dave Rogers
Partner, Clever Canary

M: 646 286 5371
http://clevercanary.com
@CleverCanary
@_DaveRogers
https://www.linkedin.com/pub/dave-rogers/0/90/92

File: app/data/files.ts [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+       WHERE f.id = $1
+       AND (
+         -- Integrated object files via component atlas
+         EXISTS (
+           SELECT 1 FROM hat.component_atlases ca 
+           WHERE f.component_atlas_id = ca.id AND ca.atlas_id = $2
+         )
+         OR
+         -- Source dataset files (with or without source study)
+         f.source_dataset_id IS NOT NULL
```
[the remark this answers is not in the document]
Comment (hunterckx) [reply app/data/files.ts:206]: On second thought, accuracy to the name would probably involve doing checks for datasets being linked to atlases (or associated via a study??), but that's not useful for what the function is actually used for

File: app/services/sns-dispatcher.ts [hunk 3]
```diff
@@ -1,5 +1,5 @@
+import { SNSMessage } from "../apis/catalog/hca-atlas-tracker/aws/entities";
 import { UnauthorizedAWSResourceError } from "../apis/catalog/hca-atlas-tracker/aws/errors";
-import { SNSMessage } from "../apis/catalog/hca-atlas-tracker/aws/schemas";
```
[the remark this answers is not in the document]
Comment (hunterckx) [reply app/services/sns-dispatcher.ts:2]: Might as well save this for that ticket

**Review by hunterckx — APPROVED** [review]
Alright, I think this could be merged now!

A couple things I want to note about my changes:
- I made it so that the bionetwork value gets validated when parsing the S3 key; this should only affect the types and the error message, since an invalid bionetwork would already produce an error due to not matching any atlases
- There was a test in `atlas-lookup.test.ts` that claimed to test that the function would "prioritize exact version match over decimal-stripped version" but didn't actually test much of anything; I removed it since it was meaningless as it was and I'm doubtful that prioritization is something we need, but I'm not 100% certain on the latter point

Remaining comments of mine that I would consider to be outstanding issues (but not ones that need to be addressed now):
- Do we need both the `httpGet` function wrapping Ky and the Ky mock? (Which one should stay if not?)
  - Created #797 to address this if we want to go ahead and merge
- `confirmFileExistsOnAtlas` only being accurate to integrated objects; if we're switching back to `component_atlases` IDs for integrated objects though, that function should become obsolete anyway

**Comment by NoopDog** [comment]
Thanks so much @hunterckx
